### PR TITLE
STM32 G4 ITM support and sample

### PR DESCRIPTION
**Pull request by andysan**
Add support for the Arm Instrumentation and Trace Macrocel (ITM) in the G4 SoC definitions. This has been tested on a ST Nucleo G431RB. This change also includes a configuration fragment that can be used to test SWO using the `samples/subsys/logging/logger` example.

This change can be tested with:
```
west build -b nucleo_g431rb zephyr/samples/subsys/logging/logger/ -DEXTRA_CONF_FILE=arm_itm_swo.conf
```
After flashing the board, connect a suitable debug probe and use the SWO viewer set to 2MHz (note that the core frequency will need to be specified, 170MHz on this particular board).

### CODE REVIEW

File: soc/arm/st_stm32/stm32g4/Kconfig.series [hunk 1]
```diff
@@ -14,5 +14,6 @@ config SOC_SERIES_STM32G4X
 	select CPU_HAS_ARM_MPU
 	select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL
 	select HAS_PM
+	select HAS_SWO
```
Comment (erwango) [inline soc/arm/st_stm32/stm32g4/Kconfig.series:17]: Out of curiosity, what is the use of `HAS_SWO`. I can't find any related code

**Review by erwango — COMMENTED** [review]
Thanks for this work, but I can't find related code in tree. Do you have any pointers or code sample that I can look at ?

File: soc/arm/st_stm32/stm32g4/Kconfig.series (hunk 1 above)
Comment (andysan) [reply soc/arm/st_stm32/stm32g4/Kconfig.series:17]: It's only used in `subsys/logging/backends/Kconfig.swo` as a dependency for `LOG_BACKEND_SWO`.

Comment (erwango) [reply soc/arm/st_stm32/stm32g4/Kconfig.series:17]: Ok, my current working branch is not recent enough. I'll have a look once I'm done.

**Comment by andysan** [comment]
> Thanks for this work, but I can't find related code in tree. Do you have any pointers or code sample that I can look at ?

There isn't exactly an example in the tree. I have been using `samples/subsys/logging/logger/` and just setting `CONFIG_LOG_BACKEND_SWO=y` and then `CONFIG_LOG_BACKEND_SWO_FREQ_HZ=2000000` to match Cube Programmer's default frequency. You'll also need to set LOG_BACKEND_SWO_REF_FREQ_HZ if it hasn't already picked up its default from the CPU node or the `swo-ref-frequency` property in the ITM DT.

Looking at my own comment, I think there are a few more changes that I need to add to  this PR:

1. A commit that sets `swo-ref-frequency` for the board(s) I have tested if needed.
2. A commit that sets the default SWO output freq to 2MHz for the boards I have tested to ensure they match the defaults in the Cube programmer.
3. A commit that adds an `arm_itm_swo.conf` config overlay to the logger example. 

As far as I can tell, there is no separate pinmux configuration needed to use the SWO port.

Does this sounds like a good plan?

**Comment by erwango** [comment]
> Does this sounds like a good plan?

That would be great.

**Comment by andysan** [comment]
It's worth noting that a device tree node isn't strictly necessary to use SWO. The log backend currently uses the `HAS_SWO` config option to detect if ITM is available. The configuration system then tries to determine the frequency of the clock driving the ITM block by first looking for the ITM node, then the clock node of the CPU. It's possible, even likely, that the device tree updates in this change aren't actually needed.

**Comment by andysan** [comment]
> > Does this sounds like a good plan?
> 
> That would be great.

OK, I have dropped the DT changes since they weren't needed. It feels a bit weird since that's how we usually detect devices, but SWO is apparently an exception. I have also added a configuration fragment that can be used for testing. The example defaults to 2MHz to ensure a well-defined frequency.

**Review by erwango — APPROVED** [review]
Ok for the STM32 part.
I let logging maintainer comment on the sample change

**Comment by erwango** [comment]
@andysan Can you check at errors in CI? They might not be directly related to you change, but they will block it and should be fixed somehow. You can raise an issue if you don't have any clues on how to fix.

**Comment by andysan** [comment]
> @andysan Can you check at errors in CI? They might not be directly related to you change, but they will block it and should be fixed somehow. You can raise an issue if you don't have any clues on how to fix.

It looks like the failing tests have been fixed on the main branch now. I have rebased and updated the PR. No changes to the PR itself though.

**Review by erwango — APPROVED** [review]
Approved on STM32 part.

**Review by FRASTM — APPROVED** [review]

**Pull request merged**